File: frontend/src/api/timeline.ts
// Contact timeline API calls -- backend endpoint is T66
// (docs/fork-plan/tickets/110-T66-contact-timeline-bounded-view-and-
// explorer.md), consumed here by the T78 web explorer. GET
// /contacts/:id/timeline returns a cursor-paginated page of a contact's
// merged timeline across all six event types, filterable by ?type=
// (comma-separated subset) and ?bucket= (recency).
import { apiFetch, API_BASE_URL, getAuthHeaders, parseErrorResponse } from './client';
import { Note } from './notes';
import { Activity } from './activities';
import { ReminderCompletion } from './reminders';
import { LifeEvent } from './lifeEvents';
import { ExternalActivity } from './externalLinks';
import { Gift } from './gifts';

// The six raw timeline event types, in the canonical order backend
// models/timeline.go uses for its sort tiebreak. Hardcoded mirror of that
// list -- this codebase has no dynamic type-list endpoint (CLAUDE.md
// frontend trap #4); keep in sync with backend/models/timeline.go by hand.
export const TIMELINE_TYPES = [
  'note',
  'activity',
  'completion',
  'life_event',
  'external_activity',
  'gift',
] as const;
export type TimelineType = (typeof TIMELINE_TYPES)[number];

// Recency buckets -- mirror of backend/models/timeline.go's TimelineBucket*
// constants; the backend validates against exactly this set.
export const TIMELINE_BUCKETS = [
  'last_7_days',
  'last_30_days',
  'last_90_days',
  'this_year',
  'all',
] as const;
export type TimelineBucket = (typeof TIMELINE_BUCKETS)[number];

// One entry in the merged timeline: the raw entity under `data` (switch on
// `type` before reading it), plus the normalized date the endpoint sorts on.
// `id` is the entity's PK as a string -- the six tables mix uint and
// UUID-string PKs -- and the row key the cursor resumes from.
export interface TimelineItem {
  type: TimelineType;
  id: string;
  date: string;
  data: Note | Activity | ReminderCompletion | LifeEvent | ExternalActivity | Gift;
}

export interface TimelineResponse {
  // Serialized `[]`, never null/absent, when the page is empty.
  items: TimelineItem[];
  // Opaque resume token; empty when there are no more rows.
  next_cursor: string;
  limit: number;
}

export interface GetTimelineParams {
  contactId: string | number;
  // Empty/undefined means all six types.
  types?: TimelineType[];
  bucket?: TimelineBucket;
  cursor?: string;
  limit?: number;
}

export async function getTimeline(params: GetTimelineParams): Promise<TimelineResponse> {
  const { contactId, types, bucket, cursor, limit = 25 } = params;
  const queryParams = new URLSearchParams({ limit: limit.toString() });
  // Empty means all types; the full set is the same query as "no filter", so
  // omit it rather than sending all six tokens.
  if (types && types.length > 0 && types.length < TIMELINE_TYPES.length) {
    queryParams.append('type', types.join(','));
  }
  if (bucket && bucket !== 'all') {
    queryParams.append('bucket', bucket);
  }
  if (cursor) {
    queryParams.append('cursor', cursor);
  }

  const response = await apiFetch(
    `${API_BASE_URL}/contacts/${contactId}/timeline?${queryParams.toString()}`,
    { headers: getAuthHeaders() }
  );
  if (!response.ok) throw await parseErrorResponse(response);
  return response.json();
}

File: backend/models/timeline.go
package models

import "time"

// Timeline event types — the six raw types a contact's timeline merges
// (T66, docs/fork-plan/tickets/110-T66-contact-timeline-bounded-view-and-
// explorer.md). They mirror the data model 1:1, matching the web timeline's
// discriminated union (ContactTimeline.tsx) and the type-filter vocabulary
// fixed by the T66/T78 platform split (design decision 1: revisit only if a
// second external-activity source ever ships alongside Immich).
const (
	TimelineTypeNote             = "note"
	TimelineTypeActivity         = "activity"
	TimelineTypeCompletion       = "completion"
	TimelineTypeLifeEvent        = "life_event"
	TimelineTypeExternalActivity = "external_activity"
	TimelineTypeGift             = "gift"
)

// TimelineTypes is the complete set of timeline event types, in the canonical
// order used for the merged sort's type tiebreak and the cursor's type
// component. Any deterministic order works; this one matches the web
// timeline's own union ordering so a page boundary sorts identically
// client- and server-side.
var TimelineTypes = []string{
	TimelineTypeNote,
	TimelineTypeActivity,
	TimelineTypeCompletion,
	TimelineTypeLifeEvent,
	TimelineTypeExternalActivity,
	TimelineTypeGift,
}

// TimelineTypeRank returns the canonical position of a timeline event type in
// TimelineTypes, and false when the token is not one of the six types. The
// rank is the tiebreak for two events that share a date but belong to
// different tables.
func TimelineTypeRank(typ string) (int, bool) {
	for i, t := range TimelineTypes {
		if t == typ {
			return i, true
		}
	}
	return 0, false
}

// Timeline recency buckets — the fixed filter set T66's design decision 2
// settled ("recent-ish vs. everything", never a free-form range picker):
// last 7 / 30 / 90 days, this year, all time. The web explorer (T78) matches
// this vocabulary exactly.
const (
	TimelineBucketLast7Days  = "last_7_days"
	TimelineBucketLast30Days = "last_30_days"
	TimelineBucketLast90Days = "last_90_days"
	TimelineBucketThisYear   = "this_year"
	TimelineBucketAll        = "all"
)

// TimelineItem is one entry in a contact's merged timeline (T66): the raw
// entity under Data, plus the normalized sort key the merge and cursor are
// built on. ID is always the entity's primary key as a string — the six
// tables mix uint (gorm.Model) and UUID-string PKs, and the cursor carries
// the id as a string for exactly that reason.
//
// Data is deliberately heterogeneous (the six entity types); clients switch
// on Type before reading it.
type TimelineItem struct {
	Type string      `json:"type"`
	ID   string      `json:"id"`
	Date time.Time   `json:"date"`
	Data interface{} `json:"data"`
}
